components/kubeconfig-service/pkg/runtime/runtime_kcp.go

```diff
@@ -0,0 +1,227 @@
+			if err != nil {
+				log.Errorf("Failed to create runtime client.")
+				return err
+			}
+			startTime, err := time.Parse("2006-01-02 15:04:05 +0000 UTC", startTimeString)
+			if err != nil {
+				log.Errorf("Failed to convert start time.")
+				return err
+			}
+			endTime := startTime.Add(ExpireTime)
+			duration := time.Until(endTime)
+			go rtc.SetupTimer(duration, runtimeID)
+		}
+	}
+
+	return nil
+}
+
+func GetK8sConfig() (*restclient.Config, error) {
+	k8sConfig, err := restclient.InClusterConfig()
+	if err != nil {
+		log.Warnf("Failed to read in cluster config: %s", err.Error())
+		log.Info("Trying to initialize with local config")
+		home := homedir.HomeDir()
+		k8sConfPath := filepath.Join(home, ".kube", "config")
+		k8sConfig, err = clientcmd.BuildConfigFromFlags("", k8sConfPath)
+		if err != nil {
+			return nil, errors.Errorf("Failed to read k8s in-cluster configuration, %s", err.Error())
+		}
+	}
+	return k8sConfig, nil
+}
+
+func GetK8sClient() (kubernetes.Interface, error) {
+	k8sconfig, err := GetK8sConfig()
+	if err != nil {
+		return nil, err
+	}
+	clientset, err := kubernetes.NewForConfig(k8sconfig)
+	if err != nil {
+		return nil, errors.Errorf("Failed to create k8s core client, %s", err.Error())
+	}
+	return clientset, err
+}
+
+func (rtc *RuntimeClient) SetupTimer(duration time.Duration, runtimeID string) {
+	userID := rtc.User.ServiceAccountName
+	if duration >= 0 {
+		timer := time.AfterFunc(duration, func() {
+			log.Printf("The timer for runtime %s user %s finished.", runtimeID, userID)
+		})
+		defer timer.Stop()
+	}
+
+	//After timer, clean up SA and ConfigMap
+	err := rtc.Cleaner()
+	if err != nil {
+		log.Warnf("Failed to clean runtime %s for user %s.", runtimeID, userID)
+	}
+
+	err = rtc.UpdateConfigMap(runtimeID)
+	if err != nil {
+		log.Warnf("Failed to clean config map for runtime %s user %s", runtimeID, userID)
+	}
+}
+
+func (rtc *RuntimeClient) UpdateConfigMap(runtimeID string) error {
+	log.Info("Trying to remove expired information.")
+	userID := rtc.User.ServiceAccountName
+
+	var patches []*JsonPatchType
+	path := "/data/" + runtimeID
+	patch := &JsonPatchType{
+		Op:   "remove",
+		Path: path,
+	}
+	patches = append(patches, patch)
+	payload, err := json.Marshal(patches)
+	if err != nil {
+		return errors.Errorf("Failed to marshal patch, %s", err.Error())
+	}
+	_, err = rtc.KcpK8s.CoreV1().ConfigMaps(KcpNamespace).Patch(context.Background(), userID, types.JSONPatchType, payload, metav1.PatchOptions{})
+	if err != nil {
+		return errors.Errorf("Failed to update config map, %s", err.Error())
+	}
+	return nil
+}
+
+func (rtc *RuntimeClient) DeployConfigMap(runtimeID string, L2L3OperatiorRole string) error {
+	userID := rtc.User.ServiceAccountName
+	tenantID := rtc.User.TenantID
+
+	log.Info("Checking if the user exists")
+	startTime := time.Now().String()
+	cm, err := rtc.KcpK8s.CoreV1().ConfigMaps(KcpNamespace).Get(context.Background(), userID, metav1.GetOptions{})
+	if k8serrors.IsNotFound(err) {
```

Comment:
I think the (get + create) is a critical section where multiple concurrent goroutines (HTTP handler routines calling the endpoint handler) could in theory race for reading this configmap of the same user and creating the CM when it doesn't exist. In practice however it should not cause issues as it's doubdtful if there will be multiple concurrent kubeconfig requests in flight for the same user. Even `kcp taskrun` command queries kubeconfigs sequentially within a loop.

All further operations (addition, deletion during expiry) are executed as individual PATCH K8S API requests, which is ensured to be serialized by the K8S API so won't cause problems .

Reply:
add a global channel to avoid critical section